### DISCUSSION TITLE
External Latch Support (8 bit parallel mode)

### DISCUSSION THREAD

**Discussion by mrcodetastic** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#226] [Ideas]
Tracking potential enhancement https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-I2S-DMA/pull/180 here.

Will need to think about this as the ESP32 (2017 edition) is horrible for TX FIFO byte reordering in 8 bit mode.

Much easier on S2 where bytes won't get messed around (FIFO is actually FIFO).

Will need to implement compiler conditional code depending on MCU type (WROOM, S2/3, C3/H2).

**Comment by AlfioSaitta** [comment]
I am working on a new project and wanted to integrate your library into my hardware. I could test the Latch too, something really good could come of it.

I am currently using an ESP32-WROOM-32D module, but almost abandoned it to replace it with an ESP32-PICO-V3-02


![image](https://user-images.githubusercontent.com/10280718/137473463-44a42731-24d7-44bf-b82f-de41d981ace5.png)

**Comment by mrcodetastic** [comment]
The use of an 8-bit latch is only really of value if there is a need for:
* GPIO pins (currently this library uses about ~12, so you'll get 4 'free')
* A larger resolution display. Not sure how much memory is saved, but probably about 30%?

If this library is to implement latch support, the first thing that needs to be designed is a open source 'sheild' similar to 
https://github.com/pixelmatix/SmartMatrix/tree/master/extras/hardware/ESP32 with 8->16 bit shift registers / buffers.

Doesn't seem to be a real use case for an 8-bit shield, as it seems most people just want to drive one or two panels in a chain.

**Comment by mrweaver** [comment]
I'd definitely be interested in this.
My current project is fairly pin heavy and I ended up using the Teensy 4.1 and SmartMatrix for my prototype, but would prefer to use a ESP32 based approach.
I'll be keeping an eye on it, and may have a look at making a ESP32 board solution for my project if I get time (probably ESP32-WROOM-32E module or similar).

[_edit: spelling_]

**Comment by solhuebner** [comment]
I am running out of pins so I am interested in the latch option :)

I thought the S3 would solve my problem but it has no Bluetooth Classic… and some other issues ;)

I am currently going through a PCB design course so if I know what is needed here I can probably put something together :)

**Reply by mrcodetastic** [reply]
Can you elaborate on what these "issues" are with the S3? Anything I should be aware of for this library?

Yeah, we really should make a latch PCB. Pixel Matrix has already built this with open Gerber files I believe.

And I think some others have created PCBs with level shifters as will.

**Reply by solhuebner** [reply]
No issues related to your library. I just had some issues with IDF and some other libraries before I realized the missing Bluetooth Classic support (for wireless gamepads in my case) and switched back to the standard ESP32.

**Comment by mrcodetastic** [comment]
With S3 support now working we really should try and design a ESP32-S3 adaptor for this device that supports TWO chains of panels in parallel.

This is possible from my analysis. There's just enough pins for 16 bits / 2 x 8 bits in parallel when also supporting all the additonal pins required for QSPI.

If espressif get PSRAM support working properly then this will be a gamechanger for being able to have very large displays from this one little device.

**Reply by DarrylStrong** [reply]
I have designed a PCB that allows two chains of 16s modules to work like a 32s chain. This allows me to drive a panel of 384x64 made up of either P2.5 128x64 modules or p4 64x32 modules.

It uses the E line to switch the clock between the two outputs ports. All the other pins are common to both ports.

**Reply by board707** [reply]
As far as I understand, the time spent on two switchable chains of 16s will be exactly the same as one 32s. Then what is the point of this improvement?
